### ISSUE TITLE
"_AS3" was implicitly appended to partition name in AS3 mode

### ISSUE THREAD

**Issue by anfernee** [F5Networks/k8s-bigip-ctlr#1171]
#### Description
Running into this issue with AS3 mode in k8s-bigip-ctlr v1.13.0. 

When I specify ${BIGIP_PARTITION} as partition name to start k8s-bigip-ctlr in as3 mode, it expected the real partition name to be ${BIGIP_PARTITION}_AS3. The partition has to end with _AS3 to be able to use CIS with AS3. 

It's a big problem for flexibility, even a bigger problem to migrate. I already have a bunch of resources created ${BIGIP_PARTITION}. Now when I switch to AS3, I have to create ${BIGIP_PARTITION}_AS3, and delete all the resources in ${BIGIP_PARTITION}, then update ConfigMap for the new partition name, then restart k8s-bigip-ctlr to use `--agent=as3`. 

#### Kubernetes Version
v1.15.7

#### Controller Version
1.13.0

#### BIG-IP Version
any

#### Diagnostic Information
k8s-bigip-ctlr manifest: 
```
      containers:
      - args:
        - --credentials-directory=/etc/bigip-login
        - --bigip-partition=$(BIGIP_PARTITION)
        - --log-level=ERROR
        - --pool-member-type=nodeport
        - '--node-label-selector= '
        - --agent=as3
        command:
        - /app/bin/k8s-bigip-ctlr
```

ConfigMap:
```
apiVersion: v1
data:
  data: '{"virtualServer":{"backend":{"serviceName":"istio-ingress","servicePort":443,"healthMonitors":[{"protocol":"tcp","interval":5,"timeout":16}]},"frontend":{"virtualAddress":{"bindAddr":"21.0.100.65","port":443},"partition":"${BIGIP_PARTITION}","balance":"ratio-member","mode":"tcp"}}}'
  schema: f5schemadb://bigip-virtual-server_v0.1.7.json
kind: ConfigMap
metadata:
  annotations:
    status.virtual-server.f5.com/ip: 21.0.100.65
  labels:
    f5type: virtual-server
    gke-f5-service: istio-ingress
    gke-managed-f5: v0.2
  name: istio-ingress-443-tcp
  namespace: gke-system
  ownerReferences:
  - apiVersion: v1
    kind: Service
    name: istio-ingress
    uid: 2c0356e3-e9fd-41c8-bad5-bee18d7820e3
  resourceVersion: "1026"
  uid: a7a9c1d9-9f19-4551-9e18-a25e66520ed8
```

bigip controller log:
```
2020/01/31 22:15:43 [WARNING] Could not get config for ConfigMap: istio-ingress-443-tcp - The partition '${BIGIP_PARTITION}' in the ConfigMap does not match '${BIGIP_PARTITION}_AS3' that the controller watches for
2020/01/31 22:16:13 [ERROR] Error parsing ConfigMap gke-system_istio-ingress-443-tcp
2020/01/31 22:16:15 [ERROR] [AS3] Big-IP Responded with code: 422
```

**Comment by mdditt2000** [comment]
You dont need to use agent=AS3 when using a configmap. Only for Routes/Ingress. Reason is you specify the agent in the configmap

kind: ConfigMap
apiVersion: v1
metadata:
  name: f5-as3-declaration
  namespace: default
  labels:
    f5type: virtual-server
    **as3: "true"**
data: 

Remove the agent and just use **as3: "true"** in the configmap and you will be fine. We are removing the ${BIGIP_PARTITION}_AS3 in CIS 2.0. CRDs are coming as a alpha in CIS 2.0 which might help you. 

Examples of configmap
https://github.com/mdditt2000/kubernetes/blob/dev/cis-1-11/A1/f5-as3-configmap.yaml

**Comment by anfernee** [comment]
Thank you Mark for the updates. Glad to see this ${BIGIP_PARTITION}_AS3 will be dropped in 2.0

**Comment by mdditt2000** [comment]
@anfernee ${BIGIP_PARTITION}_AS3 is dropped for CRDs and will be dropped for everything eles in CIS 2.1. Please monitor github issue https://github.com/F5Networks/k8s-bigip-ctlr/issues/1312

**Comment by sober-wang** [comment]
> You dont need to use agent=AS3 when using a configmap. Only for Routes/Ingress. Reason is you specify the agent in the configmap
> 
> kind: ConfigMap
> apiVersion: v1
> metadata:
> name: f5-as3-declaration
> namespace: default
> labels:
> f5type: virtual-server
> **as3: "true"**
> data:
> 
> Remove the agent and just use **as3: "true"** in the configmap and you will be fine. We are removing the ${BIGIP_PARTITION}_AS3 in CIS 2.0. CRDs are coming as a alpha in CIS 2.0 which might help you.
> 
> Examples of configmap
> https://github.com/mdditt2000/kubernetes/blob/dev/cis-1-11/A1/f5-as3-configmap.yaml

The page is not found. 👇
https://github.com/mdditt2000/kubernetes/blob/dev/cis-1-11/A1/f5-as3-configmap.yaml

**Comment by mkirkevo** [comment]
Running into a similar issue to this with 2.0.0

We deployed f5networks/k8s-bigip-ctlr:2.0.0 and now we get a tenant created from the name of our
--bigip-partition=$(BIGIP_PARTITION) with _AS3 appended to it

We have been (still are) using configMaps with as3: "true" and our controller deployment does not include --agent=as3

Because of this any as3 declaration configMap add/delete/update gets ignored going forward and our logs fill with 
[ERROR] [AS3] Invalid user-defined cfgMap with name: app-f5-config and namespace app
although our configs will apply to other cluster configuration without errors

For our dev environment we have rolled back to using 1.14.0 without any issues for now.

**Comment by mdditt2000** [comment]
@mkirkevo can you please provide me with a example of your configmap you using with CIS 1.14

**Comment by mkirkevo** [comment]
I have redacted what I needed to

```yaml
---
apiVersion: v1
kind: ConfigMap
metadata:
  name: app-f5-config
  namespace: app
  labels:
    f5type: virtual-server
    as3: "true"
data:
  template: |-
    {
        "class": "AS3",
        "declaration": {
            "class": "ADC",
            "schemaVersion": "3.11.0",
            "id": "app-f5-declaration-dev",
            "label": "App F5 Config",
            "app_k8s_dev": {
                "class": "Tenant",
                "app_dev": {
                    "class": "Application",
                    "template": "generic",
                    "app_vs_dev": {
                        "class": "Service_TCP",
                        "virtualAddresses": [
                            "10.10.10.10"
                        ],
                        "pool": "app_pool_dev",
                        "virtualPort": 443
                    },
                    "app_pool_dev": {
                        "class": "Pool",
                        "loadBalancingMode": "least-connections-member",
                        "members": [{
                            "servicePort": 0,
                            "serverAddresses": [],
                            "shareNodes": true
                        }],
                        "monitors": [{
                            "use": "app_https"
                        }]
                    },
                    "app_https": {
                        "class": "Monitor",
                        "monitorType": "https",
                        "send": "GET /app/path/health HTTP/1.1\r\nHost:\r\nConnection: Close\r\n\r\n",
                        "receive": "200"
                    },
                    "app_tls_server_dev": {
                        "class": "TLS_Client",
                        "trustCA": {"bigip": "/Common/default.crt"}
                    },
                    "app_dev_ca": {
                    "class": "CA_Bundle",
                    "bundle": "REDACTED"
                    }
                }
            }
        }
    }
```

**Comment by mdditt2000** [comment]
@mkirkevo i am doing some testing. Should should only be getting app_k8s_dev tenent. No _AS3 should be applied to a user-defined configmap. 

Standby while i do additional testing. I think i might know what is going on.

**Comment by mdditt2000** [comment]
@mkirkevo mat i am only seeing just the AS3 tenent getting created

```
2020/06/12 22:31:23 [DEBUG] [AS3] Updating AS3 Template for tenant 'AS3' app 'A1' pool 'web_pool',
2020/06/12 22:31:23 [DEBUG] [AS3] Posting AS3 Declaration
2020/06/12 22:31:23 [DEBUG] [AS3] posting request to https://192.168.200.92/mgmt/shared/appsvcs/declare/
2020/06/12 22:31:26 [DEBUG] [AS3] Response from BIG-IP: code: 200 --- tenant:AS3 --- message: success
```
applied blank

[root@k8s-1-18-master blank]# kubectl apply -f f5-as3-configmap.yaml
Warning: kubectl apply should be used on resource created by either kubectl create --save-config or kubectl apply
configmap/f5-as3-declaration configured
[root@k8s-1-18-master blank]#

Logs
[kube@k8s-1-18-master root]$ kubectl logs -f k8s-bigip-ctlr-deployment-755c95c74b-zrxxn -n kube-system | grep --color=auto -i '\[as3'

2020/06/12 22:36:48 [DEBUG] [AS3] posting request to https://192.168.200.92/mgmt/shared/appsvcs/declare/
2020/06/12 22:36:51 [DEBUG] [AS3] Response from BIG-IP: code: 200 --- tenant:AS3 --- message: success

2020/06/12 22:36:51 [DEBUG] [AS3] Preparing response message to response handler
2020/06/12 22:36:51 [DEBUG] [AS3] Sent response message to response handler
2020/06/12 22:37:08 [DEBUG] [AS3] No applications declared for tenant: AS3

removing configmap and will re-apply

[root@k8s-1-18-master A1]# kubectl delete -f f5-as3-configmap.yaml
configmap "f5-as3-declaration" deleted
[root@k8s-1-18-master A1]# kubectl create -f f5-as3-configmap.yaml
configmap/f5-as3-declaration created
[root@k8s-1-18-master A1]#

2020/06/12 22:41:50 [DEBUG] [AS3] posting request to https://192.168.200.92/mgmt/shared/appsvcs/declare/
2020/06/12 22:41:52 [DEBUG] [AS3] Response from BIG-IP: code: 200 --- tenant:AS3 --- message: success

here is my configuration used 

 args:
            - "--bigip-username=$(BIGIP_USERNAME)"
            - "--bigip-password=$(BIGIP_PASSWORD)"
            - "--bigip-url=192.168.200.92"
            - "--bigip-partition=k8s"
            - "--namespace=default"
            - "--pool-member-type=cluster"
            - "--flannel-name=fl-vxlan"
            - "--log-level=DEBUG"
            - "--insecure=true"

AS3 configuration https://github.com/mdditt2000/kubernetes-1-18/blob/master/cis%201.14/user-defined-configmap/A1/f5-as3-configmap.yaml

**Comment by mkirkevo** [comment]
@mdditt2000 
maybe we can have a quick screen share on this next week and I can reproduce this step by step?

**Comment by mdditt2000** [comment]
@mkirkevo mat please email me at m.dittmer@f5.com so i can send you a zoom meeting for next week. thanks! have a great weekend

**Comment by mdditt2000** [comment]
Issue is resolved in CIS 2.1 in the following Jira CONTCNTR-1945 - [AS3 Configmaps] Remove userDefinedAS3 Deployment argument to support multiple user Defined AS3 Configmaps --- **sprint 43**

Also Jira CONTCNTR-1678  [CIS Agent] - Single partition for agent - **Done**

**Comment by mkirkevo** [comment]
thank you Mark and team, much appreciated.

**Comment by mdditt2000** [comment]
Resolved in CIS 2.1. Please test the latest version https://github.com/F5Networks/k8s-bigip-ctlr/releases/tag/v2.1.0

Closing issue
